tests/test_llama3_prompt_creation.py

```diff
@@ -0,0 +1,172 @@
+import json
+import os
+import re
+import unittest
+from typing import List
+
+from transformers import AutoTokenizer
+
+from functionary.prompt_template import get_prompt_template_by_version
+from functionary.train.custom_datasets import prepare_training_inputs
+
+
+def extract_unmasked_chunks(labels: List[int]) -> List[List[int]]:
+    """This function is used to extract unmasked chunks of integer
+    For example, labels = [-100, -100, 1, 2, 3, -100, -100, 4, 5] --> chunks = [[1,2,3], [4,5]]
+    Args:
+        labels (List[int]): list of integer containing token_id and -100
+
+    Returns:
+        List[List[int]]: list of chunk, for example: [[1,2,3], [4,5]]
+    """
+    chunks = []
+    chunk = []
+    for token_id in labels:
+        if token_id != -100:
+            chunk.append(token_id)
+        else:
+            if len(chunk) > 0:
+                chunks.append(chunk)
+                chunk = []
+    if len(chunk) > 0:
+        chunks.append(chunk)
+    return chunks
+
+
+class TestLlama3Template(unittest.TestCase):
+    def __init__(self, *args, **kwargs):
+        super(TestLlama3Template, self).__init__(*args, **kwargs)
+
+        self.template_version = "v2.llama3"
+        self.prompt_template = get_prompt_template_by_version(self.template_version)
+
+        current_folder = os.path.dirname(os.path.abspath(__file__))
+        with open(
+            os.path.join(current_folder, f"test_case_{self.template_version}.json")
+        ) as f:
+            self.test_case = json.loads(f.read())
+
+        with open(
+            os.path.join(current_folder, f"prompt_test_{self.template_version}.txt")
+        ) as f:
+            self.final_prompt = f.read().strip()
+
+    def test_final_prompt_generation(self):
+        tools_or_functions = (
+            self.test_case["tools"]
+            if "tools" in self.test_case
+            else self.test_case["functions"]
+        )
+        final_prompt = self.prompt_template.get_prompt_from_messages(
+            self.test_case["messages"], tools_or_functions
+        )
+        # print("--------------PROMPT-------")
+        # print(final_prompt)
+        # print("-------------")
+        self.assertEqual(
+            final_prompt,
+            self.final_prompt,
+            "wrong final prompt from: get_prompt_from_messages",
+        )
+
+    def test_prepare_training_inputs_normal_tokenizer(self):
+        """this function is used to test function: prepare_training_inputs"""
+        tokenizer = AutoTokenizer.from_pretrained(
+            "gradientai/Llama-3-8B-Instruct-Gradient-1048k", legacy=True
```

Comment:
can you use the original repo

Reply:
To use the original repo we need to pass the key, so I used a public one